### ISSUE TITLE
Create smooth normals respecting edges

### ISSUE THREAD

**Issue by jonnius** [assimp/assimp#1713]
It would be nice, if the normal generator would have an option to respect "real" edges. Of course there is always an edge between two triangles, because of the discrete nature of a mesh. But if the angle between neighbour triangles is 90° or above, the normals should not be smoothly change from one triangle to the next one. Maybe there could be a threshold to detect edges. Whenever the angle between two neighbour triangles is above the threshold, the edge between them is considered a "real" edge.

This implies that the vertices at the affected edge can not be joined (if this flag is set also), because the vertex normals at the same position have to differ. To "repair" the neighbourhood, we could insert "virtual" triangles with zero width at the edge connecting the neighbours.

**Comment by mesilliac** [comment]
I think this is already a feature - the `PP_GSN_MAX_SMOOTHING_ANGLE` importer property should be able to choose the maximum angle for smoothing normals when using `aiProces_GenSmoothNormals`.

In C++:
```Assimp::Importer importer;
importer.SetPropertyFloat("PP_GSN_MAX_SMOOTHING_ANGLE", 80);
const aiScene* scene = importer.ReadFile(filename, aiProcess_GenSmoothNormals | other_flags );
```

**Comment by jonnius** [comment]
Awesome! It works well. Thanks!

**Comment by jonnius** [comment]
The current implementation works quite beside. The only negative side effect is that duplicate vertices are not joined any more. Obviously this makes sense because the curvature is a discontinuity in curvature on edges. Unfortunately, this causes problems for me, because neighbourhood information gets lost. 

Would it be possible to add an optional slightly different approach to fix this? I would propose to add additional triangles along edges that connect those vertices that are identical besides there normal.

**Comment by mesilliac** [comment]
i actually have had a similar problem with vertices in assimp - i would like a storage option that preserves the vertex positions separately from normal / uv data etc.

Adding triangles would cause a rendering burden (because those extra triangles would have to be rendered), so i think solving the problem another way would be better.

How are you using / constructing the neighbourhood data now?

**Comment by jonnius** [comment]
I did not understand what you meant about preserving the vertex position? 

I am doing some calculation on the 3d model, not actually rendering it. I create a list with the same size as the vertices list size. Each component is a list of indices pointing at the triangles the vertices is connected to. I am iterating the faces, adding the face index to each of the three vertices face list. Thus, each face knows the related vertices and each vertex knows the related faces. This way I can "walk" over the surface.

**Comment by mesilliac** [comment]
Ahh, okay i see what you mean. And how adding faces would solve it.

For the vertex positions - assimp has one index `i` for each face-vertex which corresponds to a unique combo of position, normal, uv, etc. So position is `mVertices[i]`, normal is `mNormals[i]`, etc. If in stead there could be separate indices `i, j, k` for position, normal, uv, then the position index wouldn't change even when the normals were split, and it could still be used for constructing neighbourhood data like you're doing.

Splitting positions out would also solve some related problems i have doing model processing with assimp, whereas adding faces would be problematic in my use case because i have two models that need to match. Sometimes a normal gets split on one model but not on the other, causing huge problems because the vertex count differs.

In the end i had to discard normal data entirely for the processing, and generate normals again on export.

In any case, i don't really know how to resolve either of these approaches (adding faces or splitting indices) with how assimp works now, which seems set up entirely for rendering (not model processing) :/

**Comment by jonnius** [comment]
I like your idea as it seems to fit to more use cases and mine as well. Though I am afraid your approach would need a lot of change to ASSIMP and maybe even cause challenges to provide backward compatibility. In contrast my suggestion would just be a minor change, adding an additional option from the users point of view. If ASSIMP does only focus on loading and processing models for rendering, maybe a fork would be necessary with changes to address model processing needs (not that I have the time or knowledge to do so).

**Comment by mesilliac** [comment]
Yeah you're exactly right. I think it would be worth adding an option like you say.

**Comment by tellypresence** [comment]
@jonnius Is this old feature request still valid?

**Comment by jonnius** [comment]
This feature request is no longer relevant for me. Feel free to close it.
